// Author: Alexander Thomson <thomson@cs.yale.edu>
//
syntax = "proto2";
message CalvinFSConfig {
  optional uint64 block_replication_factor = 1;
  optional uint64 metadata_replication_factor = 2;
  optional uint64 blucket_count = 3;
  optional uint64 metadata_shard_count = 4;

  // Mapping of machines to replicas.
  message ReplicaParticipant {
    optional uint64 machine = 1;
    optional uint64 replica = 2;
  }
  repeated ReplicaParticipant replicas = 11;

  // Entry in blucket/metadata map.
  message MapEntry {
    optional uint64 id = 1;
    optional uint64 replica = 2 [default = 0];
    optional uint64 machine = 3 [default = 0];
  }

  // Map of bluckets to machines
  repeated MapEntry bluckets = 21;

  // Map of metadata store shards to machines
  repeated MapEntry metadata_shards = 31;
}

message MapProto {
  message Entry {
    optional bytes key = 1;
    optional bytes value = 2;
  }
  repeated Entry entries = 1;
}

